### PR TITLE
pkcs11: Serialization fixes

### PR DESCRIPTION
**Pull request by vesajaaskelainen**
Serialization fixes for passing non-aligned data element addresses and larger return buffers.

Client fixes: https://github.com/OP-TEE/optee_client/pull/247
And tests: https://github.com/OP-TEE/optee_test/pull/479
Relates to: https://github.com/OP-TEE/optee_os/issues/4283

### CODE REVIEW

File: ta/pkcs11/src/object.c [hunk 1]
```diff
@@ -850,29 +850,39 @@ uint32_t entry_get_attribute_value(struct pkcs11_client *client,
...
+		if (!attribute_is_exportable(&cli_head, obj)) {
+			cli_head.size = PKCS11_CK_UNAVAILABLE_INFORMATION;
+			TEE_MemMove(&cli_ref->size, &cli_head.size,
+				    sizeof(cli_head.size));
 			attr_sensitive = 1;
 			continue;
 		}
 
+		/* Get real data pointer from template data */
+		data_ptr = &cli_ref->data[0];
```
Comment (etienne-lms) [inline ta/pkcs11/src/object.c]: Prefer `data_ptr = cli_ref->data;
`

File: ta/pkcs11/src/sanitize_object.c [hunk 2]
```diff
@@ -313,11 +313,13 @@ enum pkcs11_rc sanitize_client_object(struct obj_attrs **dst, void *src,
 		    pkcs11_attr_is_boolean(cli_ref.id))
 			continue;
 
-		rc = sanitize_indirect_attr(dst, &cli_ref, data);
-		if (rc == PKCS11_CKR_OK)
-			continue;
-		if (rc != PKCS11_RV_NOT_FOUND)
-			return rc;
+		if (pkcs11_attr_has_indirect_attributes(cli_ref.id)) {
+			rc = sanitize_indirect_attr(dst, &cli_ref, data);
```
Comment (etienne-lms) [inline ta/pkcs11/src/sanitize_object.c:317]: we could change `sanitize_indirect_attr()` to return ERROR_GENRIC on invalid IDs, and remove this the test against `PKCS11_RV_NOT_FOUND` below.

**Review by etienne-lms — COMMENTED** [review]
LGTM with comments addressed but for commit "ta: pkcs11: Fix serialization handling for non-indirect attributes" where we could remove a now useless NOT_FOUND return value.

File: ta/pkcs11/src/object.c [hunk 3]
```diff
@@ -850,29 +850,39 @@ uint32_t entry_get_attribute_value(struct pkcs11_client *client,
 
 	for (; cur < end; cur += len) {
 		struct pkcs11_attribute_head *cli_ref = (void *)cur;
+		struct pkcs11_attribute_head cli_head;
+		void *data_ptr;
```
[the remark this answers is not in the document]
Comment (vesajaaskelainen) [reply ta/pkcs11/src/object.c:854]: OK

File: ta/pkcs11/src/object.c [hunk 4]
```diff
@@ -850,29 +850,39 @@ uint32_t entry_get_attribute_value(struct pkcs11_client *client,
...
 
-		len = sizeof(*cli_ref) + cli_ref->size;
+		/* Make copy if header so that is aligned properly. */
```
[the remark this answers is not in the document]
Comment (vesajaaskelainen) [reply ta/pkcs11/src/object.c:856]: OK

File: ta/pkcs11/src/object.c (hunk 1 above)
Comment (vesajaaskelainen) [reply ta/pkcs11/src/object.c]: OK

**Comment by ruchi393** [comment]
@vesajaaskelainen , thanks for these patches. I tried them at my end. These help fix some issues which I was facing with google pkcs11test earlier.

File: ta/pkcs11/src/sanitize_object.c (hunk 2 above)
Comment (etienne-lms) [reply ta/pkcs11/src/sanitize_object.c:317]: @vesajaaskelainen, what I was suggesting is illustrated in https://github.com/etienne-lms/optee_os/commit/4f4dddef109381dd36be313f15a1b1a59f3734a1. If you're fine with this added change please feel free to use it wihtout my S-o-b.

Comment (etienne-lms) [reply ta/pkcs11/src/sanitize_object.c:317]: Note  etienne-lms@4f4ddde referred right above needs some fixing (`#include <assert.h>` + s/cli_ref.id/cli_ref->id/) to be functionnal.

Comment (vesajaaskelainen) [reply ta/pkcs11/src/sanitize_object.c:317]: I'll check this out for next round of updates.

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
"ta: pkcs11: fix get attribute data alignment problem"
"ta: pkcs11: get_attribute: fix returning values into larger buffer"
"ta: pkcs11: get_attribute: fix return value when querying value size"
"ta: pkcs11: add debug string for PKCS11_CKA_KEY_GEN_MECHANISM"

File: ta/pkcs11/src/sanitize_object.c (hunk 2 above)
Comment (vesajaaskelainen) [reply ta/pkcs11/src/sanitize_object.c:307]: Merged to commits. Thanks.

**Comment by vesajaaskelainen** [comment]
Tags applied and updated commits based on feedback.

**Review by etienne-lms — COMMENTED** [review]
for commit "ta: pkcs11: Fix serialization handling for non-indirect attributes":
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by ruchi393** [comment]
`Reviewed-by: Ruchika Gupta <ruchika.gupta@linaro.org>`

**Comment by vesajaaskelainen** [comment]
Thanks for the reviews!

Applied tags and pushed.

**Pull request merged**